### PR TITLE
Add dev nvidia with nv

### PR DESCRIPTION
**Pull request by GodloveD**
**Description of the Pull Request (PR):**

Several users noted that the `--nv` option does not work in conjunction with the `--contain` option.  This is because `/dev/nvidia*` devices are not bind mounted into the container if `--contain` is used.  This PR fixes this issue by bind mounting `/dev/nvidia*` if the `--nv` option is passed along with the `--contain` option.  

I'm not sure why the commit history includes all of the other commits.  It could be because that history was inadvertently lost in the release-2.4 branch when bug fixes were backported by cherry-picking merges.  If that is the case it will actually be a good thing to add this history back into the release-2.4 branch where it belongs.

Note that this PR is slated for 2.4.4.  So it should not be merged into release-2.4 until the 2.4.3 release has been tagged.  

**This fixes or addresses the following GitHub issues:**

- Fixes https://github.com/singularityware/singularity/issues/1306

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [x] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [ ] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [ ] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

**Comment by GodloveD** [comment]
@ctmadison and @paulo7777 this should be ready for testing.

**Comment by dtrudg** [comment]
Confirm this is working for me. GPU visible with `--nv` and `--contain`

```
$ singularity shell --nv --contain digits-6.0.simg

>>> from tensorflow.python.client import device_lib
>>> local_device_protos = device_lib.list_local_devices()
2018-03-09 15:15:56.051745: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2018-03-09 15:15:56.051784: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2018-03-09 15:15:56.051798: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2018-03-09 15:15:56.154121: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:893] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-03-09 15:15:56.154419: I tensorflow/core/common_runtime/gpu/gpu_device.cc:940] Found device 0 with properties:
name: GeForce GTX 960
major: 5 minor: 2 memoryClockRate (GHz) 1.2785
pciBusID 0000:01:00.0
Total memory: 3.95GiB
Free memory: 3.74GiB
2018-03-09 15:15:56.154437: I tensorflow/core/common_runtime/gpu/gpu_device.cc:961] DMA: 0
2018-03-09 15:15:56.154444: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   Y
2018-03-09 15:15:56.154464: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 960, pci bus id: 0000:01:00.0)
```

**Review by dtrudg — APPROVED** [review]
Tested (see comment), and looks good to me.

On the C front - only thing I see would be if `singularity_registry_get("NVDEV")` could return an incorrectly terminated string, but I don't believe that's the case.

**Comment by GodloveD** [comment]
Thanks for merging the CHANGELOG.md @dctrud !

File: src/lib/runtime/mounts/dev/dev.c [hunk 1]
```diff
@@ -103,6 +105,17 @@ int _singularity_runtime_mount_dev(void) {
         bind_dev(sessiondir, "/dev/random");
         bind_dev(sessiondir, "/dev/urandom");
 
+        /* if the user passed the --nv flag and the --contain flag, still bind
+        nvidia devices */
+        if ( nvdevs != NULL ) {
+            char *nvdev = strtok(nvdevs, ",");
+            while ( nvdev != NULL ) {
+                singularity_message(2, "Binding device %s\n", nvdev);
+                bind_dev(sessiondir, nvdev);
```
Comment (cclerget) [inline src/lib/runtime/mounts/dev/dev.c]: @GodloveD That's introduce a security issue, SINGULARITY_NVDEV is fully controlled by user so he could bind anything he want into container without sanity checks

**Review by cclerget — CHANGES_REQUESTED** [review]
We must ensure SINGULARITY_NVDEV contains only paths starting with /dev/

Edit: or just use opendir/readdir directly in C code when --nv flag requested and match for nvidia entries

**Comment by GodloveD** [comment]
Thanks for the review @cclerget.  And thanks for the suggestion to use opendir and readdir.  I think it makes the code cleaner and hopefully it makes the code safer as well.

**Comment by GodloveD** [comment]
As far as I know, @paulo7777 has tested this and it seems to be working for him.  @paulo7777 can you verify?

**Review by cclerget — APPROVED** [review]

**Comment by dtrudg** [comment]
@paulo7777 confirmed - also working for me here and @cclerget has approved from a security standpoint. Merging this.

**Review by dtrudg — APPROVED** [review]
Confirm this is working for me with the further changes, and Paulo has confirmed okay too.

**Comment by GodloveD** [comment]
Yay!

**Pull request merged**